feat(posts): three questions before you trust a benchmark (series 5/5) (#316)
kakkoyun/me: content/posts/go-benchmarks-lying-three-questions.md

---
title: "Three Questions Before You Trust a Benchmark"
description: "A closing synthesis of the Why Your Go Benchmarks Are Lying series — the OPERA analogy, a real CI regression that turned out to be a speedup, and three questions you can answer with three CLIs before you merge."
date: 2026-09-29T00:00:00Z

> publishDate: 2026-09-29T00:00:00Z

Author: New slot: **Friday, September 18** (moved from Tue Sep 29), closing the series' Friday run.

```suggestion
date: 2026-09-18T00:00:00Z
publishDate: 2026-09-18T00:00:00Z
```

---
_Generated by [Claude Code](https://claude.ai/code)_

> publishDate: 2026-09-29T00:00:00Z

Author: Supersedes my earlier Sep 18 suggestion (not applied). **Fri Sep 25**, closing the series.

```suggestion
date: 2026-09-25T00:00:00Z
publishDate: 2026-09-25T00:00:00Z
```

---
_Generated by [Claude Code](https://claude.ai/code)_

```go
// The entire timed loop inside BenchmarkOTLPProtoSize:
proto.Size(tracesData)
```

That is a protobuf size computation on a struct assembled entirely before `b.ResetTimer()`. It never calls `ContextWithSpan`, `SpanFromContext`, or any code the PR modified. So the reported regression had no plausible causal path through the diff — whatever the bot was measuring, it was not the changed code doing more work. Hold that thought; the mechanism turns out to be real, and stranger.

Step two: check local variance. Running `BenchmarkOTLPProtoSize` repeatedly on the development machine gave **<0.1% run-to-run variance**. The CI signal of 6–9% was not generic runner noise on this box — it was specific to CI.

Step three: build `main` and `#4891` on the same machine and compare with `benchstat`:

| Build | 1 span | 10 spans |
|-------|--------|----------|
| main | 883.3 ns/op | 7115 ns/op |
| #4891 | 840.7 ns/op | 6775 ns/op |

**#4891 was faster.** CI had flagged a regression; the same-machine A/B showed the opposite.

> The mechanism: restructuring `context.go` shifted function addresses across the `ddtrace/tracer` package. That moved the hot `proto.Size` loop's instruction fetch window relative to cache-line and branch-target buffer boundaries. At ~390 ns per iteration, small alignment shifts produce several-percent swings in either direction — enough to flip the verdict from "improvement" to "regression" on a shared runner that cannot lock CPU frequency.

Author: Consistency check: this says "~390 ns per iteration," but the table above shows 883.3/840.7 ns/op (1 span) and ~7 µs (10 spans) — 390 doesn't correspond to either row. If 390 ns is the CI runner's per-iteration figure (vs your M-series local numbers), say so; as written the reader can't reconcile the two numbers in the same story.

---
_Generated by [Claude Code](https://claude.ai/code)_

Author: Fixed in cc65c8f. I don't have a sourced CI-specific figure to reconcile against the table, so rather than invent one I reworded to point at the ~840-883 ns/op scale shown in the table above — that's enough to support "sub-microsecond scale, small shifts move the needle" without introducing an unverifiable number. Let me know if you'd rather have the actual CI-runner figure sourced instead.

---
_Generated by [Claude Code](https://claude.ai/code)_

| # | Question | Post | What to verify |
|---|----------|------|----------------|
| 1 | Is the compiler measuring real work? | [Compiler honesty](/posts/go-benchmarks-lying-compiler-honesty/) | Sink pattern present; no discarded results; `allocs/op` > 0 when allocation is expected |
| 2 | Is my sample stable enough? | [Statistics](/posts/go-benchmarks-lying-statistics/) | CV < ~5%; `benchstat` p-value < 0.05; `-count=10` minimum |
| 3 | Is the difference large relative to the noise? | [Local reproduction](/posts/go-benchmarks-lying-local-reproduction/) & [CI](/posts/go-benchmarks-lying-ci/) | Environment diagnosed; A/B on the same controlled machine; CI used for detection, not as the primary measurement |

Each question gates the next. A benchmark the compiler has optimised away answers question 2 with noise. A noisy environment makes question 3 unanswerable regardless of sample size.

> "Not all fast software is world-class, but all world-class software is fast. Performance is _the_ killer feature."
>
> Tobi Lütke (@tobi), [X, 5 May 2024](https://x.com/tobi/status/1787139157078188180)

> A Google experiment found that a half-second increase in search result page generation time caused a 20% drop in traffic. The Google search team measured the difference and caught it before shipping. Measurement errors work in both directions: they block improvements and wave through regressions.

Author: Two problems with the Google anecdote:

1. **Unlinked** — house rule is fact-based *and linked*. The canonical source for the half-second/20% figure is Marissa Mayer's Web 2.0 2006 talk as reported by Greg Linden: http://glinden.blogspot.com/2006/11/marissa-mayer-at-web-20.html
2. **"caught it before shipping" is backwards** — the 20% traffic drop was *measured on live users* in the experiment; that's how they got the number. The sentence as written invents a happier ending and contradicts the story's own logic (you can't measure a traffic drop you prevented).

```suggestion
A Google experiment [reported by Marissa Mayer in 2006](http://glinden.blogspot.com/2006/11/marissa-mayer-at-web-20.html) found that a half-second increase in search result page generation time caused a 20% drop in traffic. They could act on it because they had measured it. Measurement errors work in both directions: they block improvements and wave through regressions.
```

---
_Generated by [Claude Code](https://claude.ai/code)_

### The Minimum Viable Discipline

If you take one thing from this series: run benchmarks ten times, not once.

```bash
# Baseline on the current branch
go test -bench=. -count=10 ./... | tee old.txt

# After your change
go test -bench=. -count=10 ./... | tee new.txt

# Compare
benchstat old.txt new.txt
```

Add all three: `honestbench -r .` before reading any numbers, `benchenv` on any new machine or CI runner, and `benchgate -cv-threshold 5.0` as a gate that fails early when the environment is too noisy to produce a reliable signal.

Three tools, under an hour to wire up, applicable to any Go project. The benchmarks you write after this series will at minimum tell you what they are measuring, and tell you when the answer cannot be trusted.

---

### Resources

- [Why Your Go Benchmarks Are Lying (And How to Stop Them)](https://github.com/kakkoyun/gopherconuk-26) — GopherCon UK 2026 talk repo, demo module, and CLIs
- [The Benchmarks That Measure Themselves Away](/posts/go-benchmarks-lying-compiler-honesty/) — post 1: dead-code elimination, inlining, the sink pattern
- [Your Benchmark's Mean Is a Lie](/posts/go-benchmarks-lying-statistics/) — post 2: distributions, `benchstat`, CV

> - [What Tene's Coordinated Omission Means for `testing.B`](/posts/go-benchmarks-lying-local-reproduction/) — post 3: `b.StopTimer()` misuse, GC pause hiding

Author: Lines 194–196 list posts 1–3 under titles that belong to an earlier outline, not the drafts in #312–#314:

- L194 "The Benchmarks That Measure Themselves Away" → actual: **"The Go Benchmark That Measured Nothing: Compiler Honesty in testing.B"**
- L195 "Your Benchmark's Mean Is a Lie" → actual: **"A Single Benchmark Number Is a Lie"**
- L196 "What Tene's Coordinated Omission Means for `testing.B`" → actual: **"Before CI: Can You Trust a Benchmark on Your Own Laptop?"** — and this one is worse than a title: the blurb ("`b.StopTimer()` misuse, GC pause hiding") describes content that isn't in post 3, and **coordinated omission is covered nowhere in the series**. A reader who arrives via Tene's talk in the next bullet will go looking for a section that doesn't exist.

The URLs are right; only titles/blurbs are stale. Sync them to the shipped titles (and either cut the Tene resource or add an honest one-line blurb like "latency-measurement pitfalls beyond this series' scope").

---
_Generated by [Claude Code](https://claude.ai/code)_
